go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
+		return oc.handlePeerNamespaceSelectorOnUpdate(extraParameters.np, extraParameters.gp, func() bool {
+			// ... on condition that the removed address set was in the 'gress policy
+			return extraParameters.gp.delNamespaceAddressSet(namespace.Name)
+		})
+
+	case factory.PeerPodForNamespaceAndPodSelectorType:
+		extraParameters, ok := objectsToRetry.extraParameters.(*NetworkPolicyExtraParameters)
+		if !ok {
+			return fmt.Errorf("could not cast extraParameters to " +
+				" *NetworkPolicyExtraParameters")
+		}
+		return oc.handleLocalPodSelectorDelFunc(
+			extraParameters.policy,
+			extraParameters.np,
+			extraParameters.portGroupIngressDenyName,
+			extraParameters.portGroupEgressDenyName,
+			obj)
+
+	case factory.LocalPodSelectorTypeVar:
+		extraParameters, ok := objectsToRetry.extraParameters.(*NetworkPolicyExtraParameters)
+		if !ok {
+			return fmt.Errorf("could not cast extraParameters to " +
+				" *NetworkPolicyExtraParameters")
+		}
+
+		return oc.handleLocalPodSelectorDelFunc(
+			extraParameters.policy,
+			extraParameters.np,
+			extraParameters.portGroupIngressDenyName,
+			extraParameters.portGroupEgressDenyName,
+			obj)
+
+	default:
+		return fmt.Errorf("object type %v not supported", objectsToRetry.oType)
+	}
+
+	return nil
+}
+
+// iterateRetryResources checks if any outstanding resource objects exist and if so it tries to
+// re-add them. updateAll forces all objects to be attempted to be retried regardless.
+func (oc *Controller) iterateRetryResources(r *retryObjs, updateAll bool) {
+	r.retryMutex.Lock()
+	defer r.retryMutex.Unlock()
+	now := time.Now()
+	for objKey, entry := range r.entries {
+		if entry.ignore {
+			continue
+		}
+		// check if we need to create the resource object
+		if entry.newObj != nil {
+			// get the latest version of the resource object from the informer;
+			// if it doesn't exist we are not going to create the new object.
+			obj, err := oc.getResourceFromInformerCache(r.oType, objKey)
+			if err != nil && kerrors.IsNotFound(err) {
+				klog.Infof("%v %s not found in the informers cache,"+
+					" not going to retry object create", r.oType, objKey)
+				// QUESTION: shouldn't we actually delete the object?
+				entry.newObj = nil
+			} else { // TODO what if it's any other error?
+				entry.newObj = obj
+			}
+		}
+
+		entry.backoffSec = entry.backoffSec * 2
+		if entry.backoffSec > 60 {
+			entry.backoffSec = 60
+		}
+		backoff := (entry.backoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		objTimer := entry.timeStamp.Add(backoff)
+		if !updateAll && now.Before(objTimer) {
+			klog.Infof("%v retry %s not after timer yet, time: %s", r.oType, objKey, objTimer)
+			continue
+		}
+
+		klog.Infof("%v retry: %s retry object setup", r.oType, objKey)
+
+		// delete old object if needed
+		if entry.oldObj != nil {
+			klog.Infof("%v retry: removing old object for %s", r.oType, objKey)
+			if !isResourceScheduled(r.oType, entry.oldObj) {
```

Comment:
do we need to check is resource scheduled for an object to be deleted? I guess its just to prevent an unnecessary delete call?

Reply:
I suppose so. This check was originally done for all pods to retry, whether we had to add them or delete them: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/pods_retry.go#L41-L44
I moved it here and also further below to the "add" block, just because in this check we need to inspect the kapi structure, which is stored in `oldObj` and `newObj` respectively.